functions/spanner/src/test/java/functions/HelloSpannerTest.java

```diff
@@ -0,0 +1,156 @@
+  private static final TestLogHandler logHandler = new TestLogHandler();
+  private static boolean originalUseParentHandlers;
+
+  private BufferedWriter writerOut;
+  private StringWriter responseOut;
+
+  @BeforeClass
+  public static void setupTestLogging() {
+    LOGGER.addHandler(logHandler);
+    originalUseParentHandlers = LOGGER.getUseParentHandlers();
+    LOGGER.setUseParentHandlers(false);
+  }
+
+  @AfterClass
+  public static void restoreLogging() {
+    LOGGER.removeHandler(logHandler);
+    LOGGER.setUseParentHandlers(originalUseParentHandlers);
+  }
+
+  @Before
+  public void beforeTest() throws IOException {
+    Mockito.mockitoSession().initMocks(this);
+
+    request = PowerMockito.mock(HttpRequest.class);
+    response = PowerMockito.mock(HttpResponse.class);
+    client = PowerMockito.mock(DatabaseClient.class);
+
+    BufferedReader reader = new BufferedReader(new StringReader("{}"));
+    PowerMockito.when(request.getReader()).thenReturn(reader);
+
+    responseOut = new StringWriter();
+    writerOut = new BufferedWriter(responseOut);
+    PowerMockito.when(response.getWriter()).thenReturn(writerOut);
+
+    logHandler.clear();
+  }
+
+  private void setupSuccessfulMockQuery() {
+    ReadContext readContext = PowerMockito.mock(ReadContext.class);
+    ResultSet resultSet = PowerMockito.mock(ResultSet.class);
+    PowerMockito.when(resultSet.next()).thenReturn(true, true, false);
+    PowerMockito.when(resultSet.getLong("SingerId")).thenReturn(1L, 2L, 0L);
+    PowerMockito.when(resultSet.getLong("AlbumId")).thenReturn(1L, 1L, 0L);
+    PowerMockito.when(resultSet.getString("AlbumTitle")).thenReturn("Album 1", "Album 2", null);
+    PowerMockito.when(
+            readContext.executeQuery(
+                Statement.of("SELECT SingerId, AlbumId, AlbumTitle FROM Albums")))
+        .thenReturn(resultSet);
+    PowerMockito.when(client.singleUse()).thenReturn(readContext);
+  }
+
+  @Test
+  public void functionsHelloSpanner_shouldListAlbums() throws Exception {
+    setupSuccessfulMockQuery();
+    new HelloSpanner() {
+      @Override
+      DatabaseClient getClient() {
+        return client;
+      }
+    }.service(request, response);
+    writerOut.flush();
+    assertThat(responseOut.toString()).isEqualTo("Albums:\n1 1 Album 1\n2 1 Album 2\n");
+  }
+
+  private void setupFailedMockQuery() {
+    ReadContext readContext = PowerMockito.mock(ReadContext.class);
+    PowerMockito.when(
+            readContext.executeQuery(
+                Statement.of("SELECT SingerId, AlbumId, AlbumTitle FROM Albums")))
+        .thenThrow(
+            SpannerExceptionFactory.newSpannerException(
+                ErrorCode.NOT_FOUND, "Table `Albums` not found"));
+    PowerMockito.when(client.singleUse()).thenReturn(readContext);
+  }
+
+  @Test
+  public void functionsHelloSpanner_shouldShowQueryError() throws Exception {
+    setupFailedMockQuery();
+    new HelloSpanner() {
```

Comment:
Should this be moved to a helper function, or perhaps its own class?

Reply:
No - this is a very common idiom for Java 7.  In an ideal world, we be using Java 8 and lambda's.

Reply:
Functions is Java 8 and 11 only - so do we need to worry about Java 7?

Reply:
I don't think re-writing this into using a lambda would make it more readable. That would mean adding a `Supplier<DatabaseClient>` field to `HelloSpanner` that can be set from the constructor. So I feel it would clutter the `HelloSpanner` class with even more code only needed for testing than the current solution.